vllm/core/scheduler.py

```diff
@@ -763,6 +763,84 @@ def _get_prompt_limit(self, seq_group: SequenceGroup) -> int:
         else:
             return prompt_limit
 
+    def _get_priority(self,
+                      seq_group: SequenceGroup) -> Tuple[Optional[int], float]:
+        """ Get the priority of the sequence group.
+        Highest preference to user-defined priority, followed by arrival time.
+        Args:
+            seq_group: The sequence group input.
+        Returns:
+            The priority of the sequence group.
+        """
+        return (seq_group.priority, seq_group.arrival_time)
+
+    def _schedule_priority_preemption(
+        self,
+        budget: SchedulingBudget,
+    ) -> int:
+        """Sorts waiting and running queue. Also, force preempt requests
+        from the running queue if their priority is lower.
+        Priority-based preemption is used with the priority policy.
+        Args:
+            budget: The scheduling budget. The argument is in-place updated
+                when any requests are scheduled.
+        Returns:
+            A count of priority-based preemptions.
+        """
+
+        running_queue = self.running
+        waiting_queue = self.waiting
+
+        running_queue = deque(
+            sorted(running_queue, key=lambda x: self._get_priority(x)))
+
+        blocks_to_swap_out: List[Tuple[int, int]] = []
+        force_preemption_cnt = 0
+
+        if waiting_queue:
+            seq_group = waiting_queue[0]
+            waiting_queue.popleft()
+            num_new_seqs = seq_group.get_max_num_running_seqs()
+            num_new_tokens = self._get_num_new_tokens(seq_group,
+                                                      SequenceStatus.WAITING,
+                                                      False, budget)
+
+            # Only preempt if priority inversion exists
+            while running_queue and self._get_priority(
+                    running_queue[-1]) > self._get_priority(seq_group):
+                #Only preempt if waiting sequence cannot be allocated
+                can_allocate = self.block_manager.can_allocate(seq_group)
+                if (num_new_tokens > 0
+                        and budget.can_schedule(num_new_tokens=num_new_tokens,
+                                                num_new_seqs=num_new_seqs)
+                        and can_allocate == AllocStatus.OK):
```

Comment:
reorder for minor optimization
```suggestion
                if (num_new_tokens and can_allocate == AllocStatus.OK 
                        and budget.can_schedule(num_new_tokens=num_new_tokens,
                                                num_new_seqs=num_new_seqs)):
```